packages/shared/src/components/buttons/ButtonV2.tsx

```diff
@@ -0,0 +1,256 @@
+import React, {
+  forwardRef,
+  HTMLAttributes,
+  ReactElement,
+  ReactNode,
+  Ref,
+} from 'react';
+import classNames from 'classnames';
+import { IconProps, IconSize } from '../Icon';
+import { Loader } from '../Loader';
+import { combinedClicks } from '../../lib/click';
+
+import { ColorName as ButtonColor } from '../../styles/colors';
+
+export enum ButtonSize {
+  XLarge = 'xlarge', // only used for iconOnly buttons
+  Large = 'large',
+  Medium = 'medium',
+  Small = 'small',
+  XSmall = 'xsmall',
+}
+
+const buttonSizeToIconSize: Record<ButtonSize, IconSize> = {
+  [ButtonSize.XLarge]: IconSize.XLarge,
+  [ButtonSize.Large]: IconSize.Large,
+  [ButtonSize.Medium]: IconSize.Medium,
+  [ButtonSize.Small]: IconSize.Small,
+  [ButtonSize.XSmall]: IconSize.XSmall,
+};
+
+export type IconType = React.ReactElement<IconProps>;
+
+export { ButtonColor };
+
+export enum ButtonVariant {
+  Primary = 'primary',
+  Secondary = 'secondary',
+  Tertiary = 'tertiary',
+  Float = 'tertiaryFloat',
+}
+
+export enum ButtonIconPosition {
+  Left = 'left',
+  Right = 'right',
+}
+interface CommonButtonProps {
+  size?: ButtonSize;
+  loading?: boolean;
+  pressed?: boolean;
+  disabled?: boolean;
+  children?: ReactNode;
+  tag?: React.ElementType & AllowedTags;
+}
+
+// when color is present, variant is required
+type ColorButtonProps =
+  | { color: ButtonColor; variant: ButtonVariant }
+  | { color?: never; variant?: ButtonVariant };
+
+// when iconPosition is present, icon is required
+type IconButtonProps =
+  | { iconPosition: ButtonIconPosition; icon: IconType }
+  | { iconPosition?: never; icon?: IconType };
+
+type BaseButtonProps = CommonButtonProps & ColorButtonProps & IconButtonProps;
+
+const useGetIconWithSize = (
+  size: ButtonSize,
+  iconOnly: boolean,
+  iconPosition: ButtonIconPosition,
+) => {
+  return (icon: React.ReactElement<IconProps>) =>
+    React.cloneElement(icon, {
+      size: icon.props?.size ?? buttonSizeToIconSize[size],
+      className: classNames(
+        icon.props.className,
+        !iconOnly && 'text-base !w-6 !h-6',
+        !iconOnly && iconPosition === ButtonIconPosition.Left && '-ml-2 mr-1',
+        !iconOnly && iconPosition === ButtonIconPosition.Right && 'ml-1 -mr-2',
+        // add margin to social icons
+        icon.props.className?.split(' ').includes('socialIcon') && 'mr-3',
```

Comment:
I feel this could be on the implementation itself no as a extra component?

Reply:
Sorry, I don't think I understand what you meant here :)

Reply:
OK, removed this special casing, we can tackle the styling properly when we apply ButtonV2 to social login buttons.